File: calc_salary.js
const fs = require('fs');
const path = require('path');

const DAYS_IN_MONTH = 31; // มีนาคม 2026
const OT_RATE = 40;

const EMP_NAMES = {
  '1':'JET','2':'Mild','3':'Mameaw','4':'Bird','5':'Ploysai',
  '6':'Chompoo','7':'Mint','8':'Looknum','9':'Aum','10':'Pluem',
  '11':'Jay','12':'Cream','13':'Jui','14':'Noi','15':'Wave',
  '16':'Amp','17':'Ple','18':'YUNG','19':'TAE','20':'Ple FB',
  '21':'TUKTA','22':'Jeab','23':'ManU'
};

function parseH(s) {
  if (!s || s.trim() === '' || s.trim() === '-') return null;
  try {
    const parts = s.trim().split(':');
    return parseInt(parts[0]) + parseInt(parts[1]) / 60.0;
  } catch(e) { return null; }
}

function roundHalf(rawH) {
  if (rawH <= 0) return 0.0;
  const lower = Math.floor(rawH * 2) / 2.0;
  const upper = lower + 0.5;
  const minsToUpper = (upper - rawH) * 60;
  return (Math.round(minsToUpper * 10000) / 10000) <= 10 ? upper : lower;
}

function isSunday(dateStr) {
  try {
    return new Date(dateStr.trim()).getDay() === 0;
  } catch(e) { return false; }
}

function kitchenLateDeduct(lateMin, dailyRate, hrsDay = 12) {
  if (lateMin <= 30) return 0.0;
  lateMin = Math.round(lateMin);
  let rph = dailyRate / hrsDay;
  let lateH = Math.floor(lateMin / 60);
  let mm = Math.floor(lateMin % 60);
  let blocks = 0;
  if (lateH === 0) {
      if (mm <= 30) blocks = 0;
      else if (mm <= 50) blocks = 1;
      else blocks = 2;
  } else {
      if (mm <= 19) blocks = lateH * 2;
      else if (mm <= 49) blocks = (lateH * 2) + 1;
      else blocks = (lateH * 2) + 2;
  }
  return Math.round(blocks * 0.5 * rph * 100) / 100;
}

function makeSr(empType, daily, monthly, shiftStart, defaultOut, hrsDay, otType, lateType=null) {
  return { empType, daily, monthly, shiftStart, defaultOut, hrsDay, otType, lateType };
}

const SHIFT = {
  '1':  makeSr('รายวัน',   380,  0,    14.0, 24.0, 10, 'football', 'football'),
  '2':  makeSr('รายวัน',   380,  0,    10.0, 21.0, 10, 'morning'),
  '3':  makeSr('รายเดือน', 0, 11000,   10.0, 21.0, 10, 'morning'),
  '4':  makeSr('รายเดือน', 0, 11000,   10.0, 21.0, 10, 'morning'),
  '5':  makeSr('รายวัน',   380,  0,    15.0,  2.0, 10,  null),
  '6':  makeSr('รายเดือน', 0, 12000,   10.0, 20.0, 10,  null),
  '7':  makeSr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  null),
  '8':  makeSr('รายเดือน', 0, 12000,   10.0, 20.0, 10, 'looknum'),
  '9':  makeSr('รายเดือน', 0, 11000,   10.0, 21.0, 10, 'morning'),
  '10': makeSr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  null),
  '11': makeSr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  null),
  '12': makeSr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  null),
  '13': makeSr('รายวัน',   700,  0,    10.0, 22.0, 12, 'kitchen'),
  '14': makeSr('รายวัน',   700,  0,    10.0, 22.0, 12, 'kitchen'),
  '15': makeSr('รายเดือน', 0, 11000,   15.0,  2.0, 10,  null),
  '16': makeSr('รายวัน',   550,  0,    10.0, 22.0, 12, 'kitchen'),
  '17': makeSr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  null),
  '18': makeSr('รายวัน',   400,  0,    14.0, 24.0, 10, 'football', 'football'),
  '19': makeSr('รายวัน',   360,  0,    14.0, 24.0, 10, 'football', 'football'),
  '20': makeSr('รายวัน',   240,  0,    17.0, 24.0,  7,  null),
  '21': makeSr('รายวัน',   380,  0,     8.0, 17.0,  9,  null),
  '22': makeSr('รายวัน',   450,  0,    10.0, 22.0, 12, 'kitchen'),
  '23': makeSr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  null),
};

function getShift(eid, cinH) {
  let sr = Object.assign({}, SHIFT[eid]);
  if (eid === '8' && cinH !== null && cinH >= 12.0) {
    sr.shiftStart = 15.0;
    sr.defaultOut = 2.0;
    sr.otType = null;
  }
  return sr;
}

function calcOt(eid, cinH, coutH, sr) {
  const otType = sr.otType;
  if (!otType || cinH === null) return 0.0;
  let cout = (coutH !== null) ? coutH : sr.defaultOut;
  if (cout < cinH) cout += 24;
  if (otType === 'looknum') {
    if (cout <= 21.1666) return 0.0;
    return roundHalf(cout - 20.0);
  } else if (otType === 'morning' || otType === 'morning_pt') {
    if (cout <= 21.1666) return 0.0;
    return roundHalf(cout - 21.0);
  } else if (otType === 'kitchen') {
    if (cinH >= 9.0) return 0.0;
    return roundHalf(10.0 - cinH);
  } else if (otType === 'football') {
    if (cinH >= 13.0) return 0.0;
    return roundHalf(14.0 - cinH);
  }
  return 0.0;
}

function calcLate(lateMin, sr, cinH=null) {
  const { empType, otType, lateType, daily, monthly, hrsDay } = sr;
  if (otType === 'kitchen') return kitchenLateDeduct(lateMin, daily, hrsDay);
  if (lateType === 'football') {
    if (cinH === null || cinH <= 14 + 10/60) return 0.0;
    let periods = 1;
    let cur = 15.0;
    while (cur <= 20.0) {
      if (cinH > cur) {
        periods++;
        cur += 1.0;
      } else {
        break;
      }
    }
    return periods * OT_RATE;
  }
  if (empType === 'พาร์ทไทม์' || empType === 'พาร์ทไทม์_override') {
    if (lateMin <= 10) return 0;
    return Math.ceil(lateMin / 60) * 40;
  }
  if (empType === 'รายเดือน') {
    if (lateMin <= 10) return 0.0;
    const rph = (monthly / DAYS_IN_MONTH) / hrsDay;
    const lateH = roundHalf(lateMin / 60.0) || 0.5;
    return Math.round(rph * lateH * 100) / 100;
  } else {
    if (lateMin <= 10) return 0.0;
    const lateH = roundHalf(lateMin / 60.0) || 0.5;
    return Math.round(OT_RATE * lateH * 100) / 100;
  }
}

// อ่าน CSV
const csvPath = path.join(__dirname, 'บัตรเวลาทั้งหมด_20260327154545_export.csv');

// ลอง encoding ต่างๆ
let rawContent = null;
for (const enc of ['utf8', 'binary']) {
  try {
    rawContent = fs.readFileSync(csvPath, enc);
    break;
  } catch(e) {}
}

// Remove BOM
if (rawContent.charCodeAt(0) === 0xFEFF || rawContent.startsWith('\uFEFF')) {
  rawContent = rawContent.slice(1);
}

// Parse CSV อย่างง่าย (handle quoted fields)
function parseCSV(text) {
  const lines = text.split(/\r?\n/).filter(l => l.trim());
  const headers = lines[0].split(',').map(h => h.trim().replace(/^"|"$/g, ''));
  const rows = [];
  for (let i = 1; i < lines.length; i++) {
    const vals = lines[i].split(',').map(v => v.trim().replace(/^"|"$/g, ''));
    const row = {};
    headers.forEach((h, j) => row[h] = vals[j] || '');
    rows.push(row);
  }
  return { headers, rows };
}

const { headers, rows } = parseCSV(rawContent);
console.log('คอลัมน์:', headers.join(', '));
console.log('จำนวนแถว:', rows.length);

// หาชื่อคอลัมน์
const keyEid = headers.find(h => h.includes('หมายเลข') || h === 'No.' || h.toUpperCase().includes('EMP'));
const keyDate = headers.find(h => h.includes('วันที่') || h === 'Date');
const keyWF = headers.find(h => h.includes('วันทำงาน') || h === 'Work Day');
const keyLate = headers.find(h => h.includes('Late'));
const keyCin = headers.find(h => h.includes('เข้า') || h === 'Clock In' || h.includes('Check In'));
const keyCout = headers.find(h => h.includes('ออก') || h === 'Clock Out' || h.includes('Check Out'));

console.log('\nKey columns:', {keyEid, keyDate, keyWF, keyLate, keyCin, keyCout});

// จัดกลุ่มตาม employee
const empMap = {};
for (const r of rows) {
  const eid = (r[keyEid] || '').trim();
  if (eid && SHIFT[eid]) {
    if (!empMap[eid]) empMap[eid] = [];
    empMap[eid].push(r);
  }
}

console.log(`\nพบพนักงาน ${Object.keys(empMap).length} คน: ${Object.keys(empMap).sort((a,b)=>+a-+b).join(', ')}`);

// คำนวณ
console.log('\n' + '='.repeat(90));
console.log(`${'#'.padEnd(4)} ${'ชื่อ'.padEnd(12)} ${'ประเภท'.padEnd(12)} ${'วันทำงาน'.padStart(8)} ${'เงินฐาน'.padStart(12)} ${'OT'.padStart(9)} ${'หักสาย'.padStart(9)} ${'รวม'.padStart(12)}`);
console.log('='.repeat(90));

let grandTotal = 0;
const summaryRows = [];

for (const eid of Object.keys(empMap).sort((a,b)=>+a-+b)) {
  const srBase = SHIFT[eid];
  let totalBase = 0, totalOt = 0, totalDeduct = 0, workDays = 0;

  for (const d of empMap[eid]) {
    const wfVal = d[keyWF] || '0';
    const wf = parseFloat(wfVal) || 0;
    const lateM = parseFloat(d[keyLate] || '0') || 0;
    const checkin  = (d[keyCin]  || '').trim();
    const checkout = (d[keyCout] || '').trim();
    const dateStr = (d[keyDate] || '').trim();

    const cinH  = parseH(checkin);
    const coutH = parseH(checkout);

    if (cinH === null && coutH === null && wf === 0) continue;

    const sr = getShift(eid, cinH);
    let dailyRate = sr.empType === 'รายเดือน'
      ? Math.round((sr.monthly / DAYS_IN_MONTH) * 100) / 100
      : sr.daily;

    // TUKTA วันอาทิตย์
    if (eid === '21' && isSunday(dateStr)) {
      totalBase += 190;
      workDays++;
      continue;
    }

    // TUKTA ออกก่อน 17:00
    if (eid === '21' && coutH !== null && coutH < 17.0) {
      const hrsWorked = coutH - (cinH !== null ? cinH : 8.0);
      totalBase += Math.round(Math.max(0, hrsWorked) * (380/9) * 100) / 100;
      workDays++;
      continue;
    }

    // พาร์ทไทม์มาแทนกะเช้า
    if (['7','10','11','12','17','23'].includes(eid) && cinH !== null && cinH < 12.0) {
      dailyRate = 350;
      sr.otType = 'morning_pt';
    }
    
    // กฎสลับกะ (ออกก่อนกำหนด หรือเข้า 17:00 แปลงเป็น Part-Time Base 240)
    // ข้อยกเว้น: Kitchen (13,14,16,22), Football (1,18,19), แม่บ้าน (21)
    let isExempt = ['1','13','14','16','18','19','21','7','10','11','12','17','23'].includes(eid);
    let actualCout = coutH;
    if (actualCout !== null && cinH !== null && actualCout < cinH) actualCout += 24;

    if (!isExempt) {
      if (actualCout !== null && actualCout <= 18.0) {
        dailyRate = 240;
        sr.empType = 'พาร์ทไทม์_override';
      } else if (cinH !== null && cinH >= 16.5) {
        dailyRate = 240;
        sr.empType = 'พาร์ทไทม์_override';
        sr.otType = null;
      }
    }

    if (cinH === null) {
      dailyRate = 0;
    } else if (coutH === null) {
      // Nothing needed here for JS CLI since note wasn't saved, but zeroing out isn't needed right now
    }
    
    workDays++;
    totalBase += dailyRate;

    // หักสาย
    const deduct = calcLate(lateM, sr, cinH);
    totalDeduct += deduct;

    // OT
    const effCout = coutH !== null ? coutH : sr.defaultOut;
    const otH = calcOt(eid, cinH, effCout, sr);
    const otMoney = Math.round(otH * OT_RATE * 100) / 100;
    totalOt += otMoney;
  }

  const net = Math.round((totalBase - totalDeduct + totalOt) * 100) / 100;
  grandTotal += net;

  const name = EMP_NAMES[eid] || '?';
  const type = srBase.empType;
  console.log(
    `#${eid.padEnd(3)} ${name.padEnd(12)} ${type.padEnd(12)} ${String(workDays).padStart(8)} ` +
    `${totalBase.toFixed(2).padStart(12)} ${totalOt.toFixed(2).padStart(9)} ${totalDeduct.toFixed(2).padStart(9)} ${net.toFixed(2).padStart(12)}`
  );
  summaryRows.push({eid, name, type, workDays, base: totalBase, ot: totalOt, deduct: totalDeduct, net});
}

console.log('='.repeat(90));
console.log(`${'รวมทั้งหมด'.padEnd(60)} ${grandTotal.toFixed(2).padStart(12)} บาท`);

// สรุปแยกตามประเภท
const byType = {};
for (const s of summaryRows) {
  if (!byType[s.type]) byType[s.type] = 0;
  byType[s.type] += s.net;
}
console.log('\n--- สรุปตามประเภท ---');
for (const [t, total] of Object.entries(byType)) {
  console.log(`${t}: ${total.toFixed(2)} บาท`);
}

File: calc_salary.py
import csv, math
from datetime import datetime
from collections import defaultdict

DAYS_IN_MONTH = 31  # มีนาคม 2026
OT_RATE = 40

# กำหนดวันที่ต้องการอนุมัติ OT (เสมือนการกดปุ่มเลือกวัน)
# ตัวอย่าง: ให้พนักงานเบอร์ 8 (Looknum) คิด OT ได้เฉพาะวันที่ '2026-03-09' เป็นต้น
# ถ้าระบุ 'ALL' จะคิด OT ให้ทุกวันที่มีสิทธิ์
APPROVED_OT_DATES = {
    # '8': ['2026-03-09', '2026-03-16'],
    # '3': ['2026-03-13'],
}

# กำหนดวันที่อนุมัติให้พนักงานรายเดือนหยุดพักผ่อน (มีสิทธิ์ได้รับค่าแรง)
# เช่น วันที่ 2 และ 3 ของพนักงานเบอร์ 8
APPROVED_OFF_DATES = {
    '8': ['2026-03-02', '2026-03-03'],
}

# ชื่อพนักงาน
EMP_NAMES = {
    '1':'JET','2':'Mild','3':'Mameaw','4':'Bird','5':'Ploysai',
    '6':'Chompoo','7':'Mint','8':'Looknum','9':'Aum','10':'Pluem',
    '11':'Jay','12':'Cream','13':'Jui','14':'Noi','15':'Wave',
    '16':'Amp','17':'Ple','18':'YUNG','19':'TAE','20':'Ple FB',
    '21':'TUKTA','22':'Jeab','23':'ManU'
}

def parse_h(s):
    if not s or s.strip() in ('', '-'): return None
    try:
        parts = s.strip().split(':')
        h, m = int(parts[0]), int(parts[1])
        return h + m/60.0
    except: return None

def round_half(raw_h):
    if raw_h <= 0: return 0.0
    lower = int(raw_h * 2) / 2.0
    upper = lower + 0.5
    mins_to_upper = (upper - raw_h) * 60
    return upper if round(mins_to_upper, 4) <= 10 else lower

def is_sunday(date_str):
    try:
        return datetime.strptime(date_str.strip(), '%Y-%m-%d').weekday() == 6
    except:
        return False

def kitchen_late(late_min, daily_rate, hrs_divisor=12.0):
    # กฎใหม่: ผ่อนผัน 10 นาทีแรก, หักแบบรายนาที = (เงินรายวัน / ชั่วโมง / 60)
    if late_min <= 10: return 0.0
    
    # หารเรทต่อนาที และปัดทศนิยม 2 ตำแหน่งให้ตรงกับตัวอย่างที่ให้มา (0.97)
    rpm_exact = (float(daily_rate) / float(hrs_divisor)) / 60.0
    rpm = round(rpm_exact, 2)
    
    # หักตามจำนวนนาทีที่สายจริง (ไม่ต้องปัดนาที)
    return round(late_min * rpm, 2)

def make_sr(emp_type, daily, monthly, shift_start, default_out, hrs_day, ot_type, late_type=None):
    return {
        'emp_type': emp_type, 'daily': daily, 'monthly': monthly,
        'shift_start': shift_start, 'default_out': default_out,
        'hrs_day': hrs_day, 'ot_type': ot_type, 'late_type': late_type
    }

SHIFT = {
    '1':  make_sr('รายวัน',   380,  0,    14.0, 24.0, 10, 'football', 'football'),
    '2':  make_sr('รายวัน',   380,  0,    10.0, 21.0, 10, 'morning'),
    '3':  make_sr('รายเดือน', 0, 11000,   10.0, 21.0, 10, 'morning'),
    '4':  make_sr('รายเดือน', 0, 11000,   10.0, 21.0, 10, 'morning'),
    '5':  make_sr('รายวัน',   380,  0,    15.0,  2.0, 10,  None),
    '6':  make_sr('รายเดือน', 0, 12000,   10.0, 20.0, 10,  None),
    '7':  make_sr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  None),
    '8':  make_sr('รายเดือน', 0, 12000,   10.0, 20.0, 10, 'looknum'),
    '9':  make_sr('รายเดือน', 0, 11000,   10.0, 21.0, 10, 'morning'),
    '10': make_sr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  None),
    '11': make_sr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  None),
    '12': make_sr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  None),
    '13': make_sr('รายวัน',   700,  0,    10.0, 22.0, 12, 'kitchen'),
    '14': make_sr('รายวัน',   700,  0,    10.0, 22.0, 12, 'kitchen'),
    '15': make_sr('รายเดือน', 0, 11000,   15.0,  2.0, 10,  None),
    '16': make_sr('รายวัน',   550,  0,    10.0, 22.0, 12, 'kitchen'),
    '17': make_sr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  None),
    '18': make_sr('รายวัน',   400,  0,    14.0, 24.0, 10, 'football', 'football'),
    '19': make_sr('รายวัน',   360,  0,    14.0, 24.0, 10, 'football', 'football'),
    '20': make_sr('รายวัน',   240,  0,    17.0, 24.0,  7,  None),
    '21': make_sr('รายวัน',   380,  0,     8.0, 17.0,  9,  None),
    '22': make_sr('รายวัน',   450,  0,    10.0, 22.0, 12, 'kitchen'),
    '23': make_sr('พาร์ทไทม์',240,  0,    17.0, 24.0,  7,  None),
}

def get_shift(eid, cin_h):
    sr = dict(SHIFT[eid])
    if eid == '8' and cin_h is not None and cin_h >= 12.0:
        sr['shift_start'] = 15.0
        sr['default_out'] = 2.0
        sr['ot_type'] = None
    return sr

def calc_ot(eid, cin_h, cout_h, sr, date_str):
    ot_type = sr['ot_type']
    if ot_type is None: return 0.0
    if cin_h is None: return 0.0
    cout = cout_h if cout_h is not None else sr['default_out']
    if cout < cin_h: cout += 24
    if ot_type == 'looknum':
        if cout <= 21.1666: return 0.0
        return round_half(cout - 20.0)
    elif ot_type in ['morning', 'morning_pt']:
        if cout <= 21.1666: return 0.0
        return round_half(cout - 21.0)
    elif ot_type == 'kitchen':
        if cin_h >= 9.0: return 0.0
        return round_half(10.0 - cin_h)
    elif ot_type == 'football':
        if cin_h >= 13.0: return 0.0
        return round_half(14.0 - cin_h)
    return 0.0

def calc_late(late_min, sr, cin_h=None):
    if sr['ot_type'] == 'kitchen':
        daily = sr['daily']
        # ใช้ hrs_day ของกะ (12 ชม.สำหรับกะครัว)
        hrs_day = sr.get('hrs_day', 12)
        return kitchen_late(late_min, daily, hrs_divisor=hrs_day)
    emp_type = sr['emp_type']
    late_type= sr.get('late_type')
    monthly  = sr['monthly']
    hrs_day  = sr['hrs_day']

    # 4.5 กลุ่ม Office (หักรายนาทีแบบเฉพาะเจาะจง ไม่ปัดเศษ)
    if late_type == 'office':
        if late_min <= 10: return 0.0
        # เรทต่อนาที = เงินเดือน / วันในเดือน / 10 ชม. / 60 นาที ปัดทศนิยม 2 ตำแหน่ง
        rpm_exact = (float(monthly) / float(DAYS_IN_MONTH) / 10.0) / 60.0
        rpm = round(rpm_exact, 2)
        return round(late_min * rpm, 2)

    if late_type == 'football':
        if cin_h is None or cin_h <= 14 + 10/60: return 0.0
        periods = 1
        cur = 15.0
        while cur <= 20.0:
            if cin_h > cur:
                periods += 1
                cur += 1.0
            else:
                break
        return periods * OT_RATE
    # 5. พาร์ทไทม์เย็น (พาร์ทไทม์) หัก 40 บาทต่อการสาย (คิดเป็นรายชั่วโมงเต็ม ไม่ปัด 30 นาที)
    if emp_type == 'พาร์ทไทม์':
        if late_min <= 10: return 0.0
        import math
        return float(math.ceil(late_min / 60.0) * 40.0)

    if emp_type == 'รายเดือน':
        if late_min <= 10: return 0.0
        rph = (monthly / DAYS_IN_MONTH) / hrs_day
        late_h = round_half(late_min / 60.0) or 0.5
        return round(rph * late_h, 2)
    else:
        if late_min <= 10: return 0.0
        late_h = round_half(late_min / 60.0) or 0.5
        return round(OT_RATE * late_h, 2)

# อ่าน CSV
csv_path = r'C:\Users\Tmango\Desktop\เงินเดือน\บัตรเวลาทั้งหมด_20260327154545_export.csv'

for enc in ['utf-8-sig', 'utf-8', 'cp874', 'tis-620']:
    try:
        with open(csv_path, encoding=enc) as f:
            reader = csv.DictReader(f)
            rows = list(reader)
        print(f"อ่านสำเร็จด้วย encoding: {enc}")
        print("คอลัมน์:", list(rows[0].keys()) if rows else "ไม่มี")
        break
    except Exception as e:
        print(f"{enc} ล้มเหลว: {e}")

# จัดกลุ่มตาม employee
emp_map = defaultdict(list)
for r in rows:
    eid = str(r.get('หมายเลขพนักงาน', '') or r.get('No.', '') or '').strip()
    if eid and eid in SHIFT:
        emp_map[eid].append(r)
    else:
        # ลองหา key อื่น
        for k in r.keys():
            if 'หมาย' in k or 'No' in k or 'ID' in k.upper():
                val = str(r[k]).strip()
                if val and val in SHIFT:
                    emp_map[val].append(r)
                    break

print(f"\nพบพนักงาน {len(emp_map)} คน: {sorted(emp_map.keys(), key=lambda x: int(x) if x.isdigit() else 99)}")

# คำนวณทีละคน
print("\n" + "="*80)
print(f"{'#':<4} {'ชื่อ':<12} {'ประเภท':<12} {'วันทำงาน':>8} {'เงินเดือน':>12} {'OT':>8} {'หักสาย':>8} {'รวม':>12}")
print("="*80)

summary = {}
for eid in sorted(emp_map.keys(), key=lambda x: int(x) if x.isdigit() else 99):
    sr_base = SHIFT[eid]
    total_base = 0.0
    total_ot   = 0.0
    total_deduct = 0.0
    work_days  = 0

    # Create mapping of date to row for easy lookup
    days_map = {}
    for d in emp_map[eid]:
        date_str = d.get('วันที่', '') or d.get('Date', '') or ''
        if date_str:
            days_map[date_str] = d

    for day in range(1, DAYS_IN_MONTH + 1):
        date_str = f"2026-03-{day:02d}"
        d = days_map.get(date_str, {})
        
        # หาคอลัมน์
        wf_val = d.get('วันทำงาน', '') or d.get('Work Day', '') or ''
        try: wf = float(wf_val) if wf_val else 0
        except: wf = 0

        late_val = d.get('Late In(M)', '') or d.get('Late', '') or '0'
        try: late_m = float(late_val) if late_val else 0
        except: late_m = 0

        checkin  = d.get('นาฬิกาเข้า', '') or d.get('Clock In', '') or ''
        checkout = d.get('นาฬิกาออก', '') or d.get('Clock Out', '') or ''

        cin_h  = parse_h(checkin)
        cout_h = parse_h(checkout)

        if cin_h is None and cout_h is None:
            if sr_base['emp_type'] == 'รายเดือน':
                # ตรวจสอบว่าเป็นวันหยุดที่อนุมัติหรือไม่
                if eid in APPROVED_OFF_DATES and date_str in APPROVED_OFF_DATES[eid]:
                    total_base += round(sr_base['monthly']/DAYS_IN_MONTH, 2)
                    work_days += 1
            continue

        sr = get_shift(eid, cin_h)
        daily_rate = round(sr['monthly']/DAYS_IN_MONTH, 2) if sr['emp_type'] == 'รายเดือน' else sr['daily']

        # TUKTA วันอาทิตย์
        if eid == '21' and is_sunday(date_str):
            total_base += 190
            work_days += 1
            continue

        # TUKTA ออกก่อน 17:00
        if eid == '21' and cout_h is not None and cout_h < 17.0:
            hrs_worked = cout_h - (cin_h if cin_h else 8.0)
            total_base += round(max(0, hrs_worked) * (380/9), 2)
            work_days += 1
            continue

        # พาร์ทไทม์มาแทนกะเช้า
        if eid in ['7','10','11','12','17','23'] and cin_h and cin_h < 12.0:
            daily_rate = 350
            sr['ot_type'] = 'morning_pt'
            
        # กฎสลับกะ (ออกก่อนกำหนด หรือเข้า 17:00 แปลงเป็น Part-Time Base 240)
        # ข้อยกเว้น: Kitchen (13,14,16,22), Football (1,18,19), แม่บ้าน (21)
        is_exempt = eid in ['1','13','14','16','18','19','21','7','10','11','12','17','23']
        actual_cout = cout_h
        if actual_cout is not None and cin_h is not None and actual_cout < cin_h:
            actual_cout += 24

        if not is_exempt:
            if actual_cout is not None and actual_cout <= 18.0:
                daily_rate = 240
                sr['emp_type'] = 'พาร์ทไทม์_override'
            elif cin_h is not None and cin_h >= 16.5:
                # สลับมา 17:00
                daily_rate = 240
                sr['emp_type'] = 'พาร์ทไทม์_override'
                sr['ot_type'] = None

        # หักยอดกรณีลืมสแกนเข้าหรือออก (เหมือน JS: รอใส่เวลาเข้า / ออก)
        if cin_h is None:
            daily_rate = 0.0
        elif cout_h is None:
            daily_rate = 0.0

        work_days += 1
        total_base += daily_rate

        # หักสาย
        deduct = calc_late(late_m, sr, cin_h)
        total_deduct += deduct

        # OT
        eff_cout = cout_h if cout_h is not None else sr['default_out']
        ot_h = calc_ot(eid, cin_h, eff_cout, sr, date_str)
        ot_money = round(ot_h * OT_RATE, 2)
        total_ot += ot_money

    net = round(total_base - total_deduct + total_ot, 2)
    summary[eid] = {'name': EMP_NAMES.get(eid, '?'), 'work_days': work_days,
                     'base': total_base, 'ot': total_ot, 'deduct': total_deduct, 'net': net,
                     'type': sr_base['emp_type']}
    print(f"#{eid:<3} {EMP_NAMES.get(eid,'?'):<12} {sr_base['emp_type']:<12} {work_days:>8} {total_base:>12,.2f} {total_ot:>8,.2f} {total_deduct:>8,.2f} {net:>12,.2f}")

print("="*80)
total_all = sum(v['net'] for v in summary.values())
print(f"{'รวมทั้งหมด':>50} {total_all:>12,.2f} บาท")
